torch/utils/_benchmark/utils/valgrind_wrapper.py

```diff
@@ -0,0 +1,227 @@
+import os
+import re
+import shutil
+import subprocess
+import sys
+import tempfile
+import textwrap
+
+import torch
+
+
+class _ValgrindWrapper(object):
+    def __init__(self):
+        self._valgrind_available = not subprocess.run(
+            ["which", "valgrind"], capture_output=True).returncode
+
+        self._build_type = None
+        build_search = re.search("BUILD_TYPE=(.+),", torch.__config__.show())
+        if build_search is not None:
+            self._build_type = build_search.groups()[0].split(",")[0]
+
+        self._baseline_cache = {}
+
+    def _validate(self):
+        if not self._valgrind_available:
+            raise OSError("Could not find `valgrind`")
+
+        if self._build_type is None:
+            raise OSError("Could not determine BUILD_TYPE")
+
+        if self._build_type != "RelWithDebInfo":
+            raise OSError(
+                "PyTorch sould be built with REL_WITH_DEB_INFO=1 "
+                "(debug symbols) in order to produce a useful C++ call graph, "
+                f"got BUILD_TYPE={self._build_type} instead"
+            )
+
+    def collect_callgrind(self, stmt: str, setup: str, number: int, num_threads: int):
+        self._validate()
+        if number not in self._baseline_cache:
+            self._baseline_cache[(number, num_threads)] = self._invoke(
+                stmt="pass", setup="pass", number=number, num_threads=num_threads)
+
+        # We include `pass` so it matches the baseline and reduces the instruction diff.
+        fn_counts = self._invoke(
+            stmt=f"pass;{stmt}", setup=setup, number=number, num_threads=num_threads)
+
+        count_dict = {fn: count for count, fn in fn_counts}
+        for count, fn in self._baseline_cache[(number, num_threads)]:
+            count_dict.setdefault(fn, 0)
```

Comment:
nit: you can use `collections.defaultdict`

Reply:
Good point. (I refactored the logic somewhat, but use `defaultdict` in the new one.)